wp-sw-cache/lib/service-worker.js

```diff
@@ -1,103 +1,155 @@
-(function(self){
-  var CACHE_PREFIX = '__wp-sw-cache::';
+(function(self, localforage){
 
-  var CACHE_NAME = CACHE_PREFIX + '$name';
-  var CACHE_FILES = $files;
+  // Represents the name of both storage and serviceworker
+  // This shouldn't change because the SW code below is smart enough to
+  // update what's needed and delete what isn't
+  var storageKey = '$name';
 
-  var debug = $debug;
+  var wpSwCache = self.wpSwCache = {
+    // A { url: hash } object which will be used to populate cache
+    // A changed url object will ensure the SW is downloaded when URLs change
+    urls: $urls,
+    // Instance of localForage to save urls and hashes to see if anything has changed
+    storage: localforage.createInstance({ name: storageKey }),
+    // Name of the cache the plugin will use
+    cacheName: storageKey,
+    // Method to cleanse a URL before comparing and caching
+    normalizeAndAnonymize: function(request) {
+      var url = new URL(request.url);
+      if (url.origin !== location.origin) {
+        return request;
+      }
 
-  self.addEventListener('install', function(event) {
-    // Perform install step:  loading each required file into cache
-    event.waitUntil(
-      caches.open(CACHE_NAME)
-        .then(function(cache) {
-          // Add all offline dependencies to the cache
-          if (debug) {
-            console.log('[install] Caches opened, adding all core components to cache');
+      url.search = '';
+      url.fragment = '';
+      return new Request(url, {
+        method: request.method,
+        headers: request.headers,
+        mode: 'no-cors',
+        credentials: request.credentials,
+        cache: request.cache,
+        redirect: request.redirect,
+        referrer: request.referrer,
+        integrity: request.integrity
+      });
+    },
+    // Detect if a URL should be cacheable and in the desired URL list
+    shouldBeHandled: function(request) {
+      return request.method === 'GET' && (request.url in this.urls);
+    },
+    // Adds URLs to cache and localForage if the file has changed or needs to be added
+    update: function() {
+      // For every URL (file) the user wants cached...
+      return Promise.all(Object.keys(this.urls).map(url => {
+        var hash = this.urls[url];
+
+        // ... get its hash from storage ...
+        return this.storage.getItem(url).then(value => {
+          // ... and if nothing has changed, just move on to the next URL
+          if(value === hash) {
+            this.log('[update] Hash unchanged, doing nothing: ', url);
+            return Promise.resolve();
           }
-          return cache.addAll(CACHE_FILES);
-        })
-        .then(function() {
-          if (debug) {
-            console.log('[install] All required resources have been cached, we\'re good!');
+          else {
```

Comment:
Nit: I don't know what you prefer, but I usually don't use `else` after an if with a return, this way you don't need to indent the code in the else branch.